sql/sort.go

```diff
@@ -0,0 +1,224 @@
+	// below. This is the set of columns requested by the query.
+	columns := s.Columns()
+	var ordering []int
+
+	for _, o := range n.OrderBy {
+		index := 0
+
+		// Normalize the expression which has the side-effect of evaluating
+		// constant expressions and unwrapping expressions like "((a))" to "a".
+		expr, err := parser.NormalizeExpr(o.Expr)
+		if err != nil {
+			return nil, err
+		}
+
+		if qname, ok := expr.(*parser.QualifiedName); ok {
+			if len(qname.Indirect) == 0 {
+				// Look for an output column that matches the qualified name.
+				target := string(qname.Base)
+				for j, col := range columns {
+					if equalName(target, col) {
+						index = j + 1
+						break
+					}
+				}
+			}
+
+			if index == 0 {
+				// No output column matched the qualified name, so look for an existing
+				// render target that matches the column name.
+				if err := qname.NormalizeColumnName(); err != nil {
+					return nil, err
+				}
+				if qname.Table() == "" || equalName(s.desc.Alias, qname.Table()) {
+					for j, r := range s.render {
+						if qval, ok := r.(*qvalue); ok {
+							if equalName(qval.col.Name, qname.Column()) {
+								index = j + 1
+								break
+							}
+						}
+					}
+				}
+			}
+		}
+
+		if index == 0 {
+			// The order by expression matched neither an output column or an
+			// existing render target.
+			if datum, ok := expr.(parser.Datum); ok {
+				// If we evaluated to an int, use that as an index to order by.
+				i, ok := datum.(parser.DInt)
+				if !ok {
+					return nil, fmt.Errorf("invalid ORDER BY: %s", expr)
+				}
+				index = int(i)
+				if index < 1 || index > len(columns) {
+					return nil, fmt.Errorf("invalid ORDER BY index: %d not in range [1, %d]",
+						index, len(columns))
+				}
+			} else {
+				// Add a new render expression to use for ordering.
+				if err := s.addRender(parser.SelectExpr{Expr: expr}); err != nil {
+					return nil, err
+				}
+				index = len(s.columns)
+			}
+		}
+
+		if o.Direction == parser.Descending {
+			index = -index
+		}
+		ordering = append(ordering, index)
+	}
+
+	return &sortNode{columns: columns, ordering: ordering}, nil
+}
+
+type sortNode struct {
+	plan     planNode
+	columns  []string
+	ordering []int
+	values   *valuesNode
+	err      error
+}
+
+func (n *sortNode) Columns() []string {
+	return n.columns
+}
+
+func (n *sortNode) Ordering() []int {
+	return n.ordering
+}
+
+func (n *sortNode) Values() parser.DTuple {
+	// If an ordering expression was used the number of columns in each row might
+	// differ from the number of columns requested, so trim the result.
+	v := n.values.Values()
+	return v[:len(n.columns)]
+}
+
+func (n *sortNode) Next() bool {
+	if n.values == nil {
+		if !n.initValues() {
+			return false
+		}
+	}
+	return n.values.Next()
+}
+
+func (n *sortNode) Err() error {
+	return n.err
+}
+
+// wrap the supplied planNode with the sortNode if sorting is required.
+func (n *sortNode) wrap(plan planNode) planNode {
```

Comment:
could you add a TODO that mentions reverse scans? i'm not sure where that would go.